### PR TITLE
Ftr: File system service discovery 

### LINKED ISSUE #426: Align 2.7.5: FileSystemServiceDiscovery
**What would you like to be added**:

FileSystemServiceDiscovery, the implementation of ServiceDiscovery based on file system.

We won't add it into 1.5.0 milestone. We will try our best to publish this feature in 1.5.0. But do not promise that.

### PR DESCRIPTION
**Pull request by DogBaoBao**
**What this PR does**:
The implementation of ServiceDiscovery based on file system.

**Which issue(s) this PR fixes**:

Fixes #426

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by DogBaoBao** [comment]
Not finished, I'm working on it.

File: common/constant/key.go [hunk 1]
```diff
@@ -169,6 +169,10 @@ const (
 	NACOS_USERNAME               = "username"
 )
 
+const (
+	FILE_KEY = "file"
```
Comment (zouyx) [inline common/constant/key.go:173]: file_discovery_service_key？

File: config_center/file/impl.go [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+	"os"
+	"os/exec"
+	"os/user"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+)
+
+import (
```
Comment (zouyx) [inline config_center/file/impl.go:33]: second block is 3rd-party pkg，third block is internal pkg

File: config_center/dynamic_configuration.go [hunk 3]
```diff
@@ -57,6 +57,8 @@ type DynamicConfiguration interface {
 
 	// PublishConfig will publish the config with the (key, group, value) pair
 	PublishConfig(string, string, string) error
+	// PublishConfig will remove the config white the (key, group) pair
```
Comment (zouyx) [inline config_center/dynamic_configuration.go:60]: comment is wrong

File: config_center/file/listener.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+)
+
+import (
+	"github.com/apache/dubbo-go/common/extension"
+	"github.com/apache/dubbo-go/common/logger"
+	"github.com/apache/dubbo-go/config_center"
+	"github.com/apache/dubbo-go/remoting"
+)
+
+import (
```
Comment (zouyx) [inline config_center/file/listener.go:32]: sort this block

File: registry/file/service_discovery.go [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+	"github.com/apache/dubbo-go/common/extension"
+	"github.com/apache/dubbo-go/common/logger"
+	"github.com/apache/dubbo-go/config"
+	"github.com/apache/dubbo-go/config_center"
+	"github.com/apache/dubbo-go/config_center/file"
+	"github.com/apache/dubbo-go/registry"
+)
+
+import (
+	gxset "github.com/dubbogo/gost/container/set"
```
Comment (zouyx) [inline registry/file/service_discovery.go:42]: sort block

File: registry/file/service_discovery_test.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+)
+
+import (
+	"github.com/apache/dubbo-go/common/constant"
+	"github.com/apache/dubbo-go/common/extension"
+	"github.com/apache/dubbo-go/config"
+	"github.com/apache/dubbo-go/registry"
+)
+
+import (
```
Comment (zouyx) [inline registry/file/service_discovery_test.go:34]: sort block

File: common/constant/key.go (hunk 1 above)
Comment (DogBaoBao) [reply common/constant/key.go:173]: yes，the key from java:

```
file=org.apache.dubbo.registry.client.FileSystemServiceDiscovery
```

File: config_center/file/listener.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+	extension.AddCustomShutdownCallback(func() {
+		cl.watch.Close()
+	})
+
+	return cl
+}
+
+func removeCallback(lmap map[config_center.ConfigurationListener]struct{}, key string, event remoting.EventType) {
+	if len(lmap) == 0 {
+		logger.Warnf("file watch callback but configuration listener is empty, key:%s, event:%v", key, event)
```
Comment (zouyx) [inline config_center/file/listener.go:91]:
```suggestion
		logger.Warnf("file watch callback but configuration listener is empty, key:%s, event:%v", key, event)
		return
```

Should return directly here

File: config_center/file/listener.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+	}
+	for l := range lmap {
+		callback(l, key, "", event)
+	}
+}
+
+func dataChangeCallback(lmap map[config_center.ConfigurationListener]struct{}, key string, event remoting.EventType) {
+	if len(lmap) == 0 {
+		logger.Warnf("file watch callback but configuration listener is empty, key:%s, event:%v", key, event)
```
Comment (zouyx) [inline config_center/file/listener.go:100]: As above?

File: config_center/file/listener.go [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+	// make a map[your type]struct{} like set in java
+	listeners, loaded := cl.keyListeners.LoadOrStore(key, map[config_center.ConfigurationListener]struct{}{listener: {}})
+	if loaded {
+		listeners.(map[config_center.ConfigurationListener]struct{})[listener] = struct{}{}
+		cl.keyListeners.Store(key, listeners)
+	} else {
+		if err := cl.watch.Add(key); err != nil {
+			logger.Errorf("watcher add path:%s err:%v", key, err)
+		}
+	}
```
Comment (zouyx) [inline config_center/file/listener.go:134]:
```suggestion
	if loaded {
		listeners.(map[config_center.ConfigurationListener]struct{})[listener] = struct{}{}
		cl.keyListeners.Store(key, listeners)
		return
	} 
		if err := cl.watch.Add(key); err != nil {
			logger.Errorf("watcher add path:%s err:%v", key, err)
		}
	
```

File: config_center/file/listener.go [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+// RemoveListener will delete a listener if loaded
+func (cl *CacheListener) RemoveListener(key string, listener config_center.ConfigurationListener) {
+	listeners, loaded := cl.keyListeners.Load(key)
+	if loaded {
+		delete(listeners.(map[config_center.ConfigurationListener]struct{}), listener)
+		if err := cl.watch.Remove(key); err != nil {
+			logger.Errorf("watcher remove path:%s err:%v", key, err)
+		}
+	}
```
Comment (zouyx) [inline config_center/file/listener.go:145]:
```suggestion
	if !loaded {
		return
	}
	delete(listeners.(map[config_center.ConfigurationListener]struct{}), listener)
		if err := cl.watch.Remove(key); err != nil {
			logger.Errorf("watcher remove path:%s err:%v", key, err)
		}
```

File: config_center/file/listener.go [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+func getFileContent(path string) string {
+	if c, err := ioutil.ReadFile(path); err != nil {
+		logger.Errorf("read file path:%s err:%v", path, err)
+		return ""
+	} else {
+		return string(c)
+	}
```
Comment (zouyx) [inline config_center/file/listener.go:154]:
```suggestion
	} 
	return string(c)
	
```

File: registry/file/service_discovery.go [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+	sdc, ok := config.GetBaseConfig().GetServiceDiscoveries(name)
+	if !ok || sdc.Protocol != constant.FILE_KEY {
+		return nil, perrors.New("could not init the instance because the config is invalid")
+	}
+
+	if rp, err := file.Home(); err != nil {
+		return nil, perrors.WithStack(err)
+	} else {
+		fdcf := extension.GetConfigCenterFactory(constant.FILE_KEY)
+		p := path.Join(rp, ".dubbo", "registry")
```
Comment (zouyx) [inline registry/file/service_discovery.go:89]:
```suggestion
		p := path.Join(rp, ".dubbo",constant.REGISTRY_KEY)
```

File: registry/file/service_discovery.go [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+			for _, v := range sd.GetServices().Values() {
+				for _, i := range sd.GetInstances(v.(string)) {
+					// like java do nothing
+					l := &RegistryConfigurationListener{}
+					sd.dynamicConfiguration.AddListener(getServiceInstanceId(i), l, config_center.WithGroup(getServiceName(i)))
+				}
+			}
+
+			return sd, nil
+		}
```
Comment (zouyx) [inline registry/file/service_discovery.go:114]:
```suggestion
		} 
			sd := &fileSystemServiceDiscovery{
				dynamicConfiguration: *c.(*file.FileSystemDynamicConfiguration),
				rootPath:             p,
				fileMap:              make(map[string]string),
			}

			extension.AddCustomShutdownCallback(func() {
				sd.Destroy()
			})

			for _, v := range sd.GetServices().Values() {
				for _, i := range sd.GetInstances(v.(string)) {
					// like java do nothing
					l := &RegistryConfigurationListener{}
					sd.dynamicConfiguration.AddListener(getServiceInstanceId(i), l, config_center.WithGroup(getServiceName(i)))
				}
			}

			return sd, nil
		
```

File: registry/file/service_discovery.go [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+	sn := getServiceName(instance)
+	if c, err := getContent(instance); err != nil {
+		return err
+	} else {
+		if err := fssd.dynamicConfiguration.PublishConfig(id, sn, c); err != nil {
+			return perrors.WithStack(err)
+		} else {
+			fssd.fileMap[id] = fssd.dynamicConfiguration.GetPath(id, sn)
+		}
+	}
```
Comment (zouyx) [inline registry/file/service_discovery.go:154]:
```suggestion
	} 
	if err := fssd.dynamicConfiguration.PublishConfig(id, sn, c); err != nil {
		return perrors.WithStack(err)
	}
	fssd.fileMap[id] = fssd.dynamicConfiguration.GetPath(id, sn)
		
```

File: registry/file/service_discovery.go [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+	return si.GetServiceName()
+}
+
+// getContent json string
+func getContent(si registry.ServiceInstance) (string, error) {
+	if bytes, err := json.Marshal(si); err != nil {
+		return "", err
+	} else {
+		return string(bytes), nil
+	}
```
Comment (zouyx) [inline registry/file/service_discovery.go:179]:
```suggestion
	} 
	return string(bytes), nil
	
```

File: registry/file/service_discovery.go [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+// Unregister will unregister this instance from registry
+func (fssd *fileSystemServiceDiscovery) Unregister(instance registry.ServiceInstance) error {
+	id := getServiceInstanceId(instance)
+	sn := getServiceName(instance)
+	if err := fssd.dynamicConfiguration.RemoveConfig(id, sn); err != nil {
+		return err
+	} else {
+		delete(fssd.fileMap, instance.GetId())
+		return nil
+	}
```
Comment (zouyx) [inline registry/file/service_discovery.go:196]:
```suggestion
	} 
		delete(fssd.fileMap, instance.GetId())
		return nil
	
```

File: registry/file/service_discovery.go [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+}
+
+// GetInstances will return all service instances with serviceName
+func (fssd *fileSystemServiceDiscovery) GetInstances(serviceName string) []registry.ServiceInstance {
+	if set, err := fssd.dynamicConfiguration.GetConfigKeysByGroup(serviceName); err != nil {
+		logger.Errorf("[FileServiceDiscovery] Could not query the instances for service{%s}, error = err{%v} ",
+			serviceName, err)
+
+		return make([]registry.ServiceInstance, 0, 0)
+	} else {
```
Comment (zouyx) [inline registry/file/service_discovery.go:227]: remove else - -

File: registry/file/service_discovery.go (hunk 12 above)
Comment (DogBaoBao) [reply registry/file/service_discovery.go]: this may not the registry meaning the registry

File: registry/file/service_discovery.go [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+
+// fileServiceDiscovery is the implementation of service discovery based on file.
+type fileSystemServiceDiscovery struct {
+	dynamicConfiguration file.FileSystemDynamicConfiguration
+	rootPath             string
+	fileMap              map[string]string
+}
+
+func newFileSystemServiceDiscovery(name string) (registry.ServiceDiscovery, error) {
+	instance, ok := instanceMap[name]
```
Comment (zouyx) [inline registry/file/service_discovery.go:66]: where do you set instance into  instanceMap? do i miss something?

File: registry/file/service_discovery.go [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+	if ok {
+		return instance, nil
+	}
+
+	initLock.Lock()
+	defer initLock.Unlock()
```
Comment (zouyx) [inline registry/file/service_discovery.go:72]: This lock is too long, i suggest unlock it after set into instanceMap .

**Review by AlexStocks — APPROVED** [review]

**Review by zouyx — APPROVED** [review]

**Review by fangyincheng — APPROVED** [review]

**Pull request merged**